packages/client/src/perma-cache.js

```diff
@@ -0,0 +1,396 @@
+                // do not retry if abortable errors - will not succeed
+                if (ABORTABLE_ERRORS_STATUS.includes(request.status)) {
+                  throw new AbortError(e)
+                }
+                /* c8 ignore next 2 */
+                throw e
+              }
+
+              onPut && onPut(url)
+
+              return {
+                ...res,
+                url, // decoded
+              }
+            },
+            /* c8 ignore next 3 */
+            {
+              retries: maxRetries == null ? MAX_RETRIES : maxRetries,
+            }
+          )
+        } catch (error) {
+          return {
+            // @ts-ignore non error type
+            error: error.message,
+            url,
+          }
+        }
+        return cacheResult
+      })
+    )
+
+    // @ts-ignore p-settle will always handle error internally
+    return cacheResults.map((cacheResult) => cacheResult.value)
+  }
+
+  /**
+   * @param {Service} service
+   * @param {ListOptions} [options]
+   * @returns {AsyncIterable<CacheEntry>}
+   */
+  static async *list(
+    { endpoint, token, rateLimiter = globalRateLimiter },
+    { sort = 'date', order = 'asc' } = {}
+  ) {
+    const headers = PermaCache.headers(token)
+    let search = new URLSearchParams({ sort, order })
+    let nextPageUrl = new URL(`perma-cache?${search}`, endpoint)
+    do {
+      await rateLimiter()
+
+      const request = await fetch(nextPageUrl.toString(), {
+        method: 'GET',
+        headers,
+      })
+      const result = await request.json()
+      if (!request.ok) {
+        throw new Error(result.message)
+      }
+
+      // Go over next links until not provided by API anymore
+      const link = request.headers.get('link')
+      if (link) {
+        nextPageUrl = new URL(
+          link.replace('<', '').replace('>; rel="next"', ''),
+          endpoint
+        )
+      } else {
+        // @ts-ignore typescript complains about URL type cannot be undefined
+        nextPageUrl = undefined
+      }
+
+      for (const entry of result) {
+        yield entry
+      }
+    } while (nextPageUrl)
+  }
+
+  /**
+   * @param {Service} service
+   * @param {string[]} urls
+   * @param {DeleteOptions} [options]
+   * @returns {Promise<CacheDeleteResult[]>}
+   */
+  static async delete(
+    { endpoint, token, rateLimiter = globalRateLimiter },
+    urls,
+    { onDelete, maxRetries = MAX_RETRIES } = {}
+  ) {
+    areNftStorageLinkUrls(urls)
+
+    const headers = PermaCache.headers(token)
+
+    /** @type {import('p-settle').PromiseResult<CacheDeleteResult>[]} */
+    const cacheDeleteResults = await pSettle(
+      urls.map(async (url) => {
+        const apiUrl = new URL(
+          `perma-cache/${encodeURIComponent(url)}`,
+          endpoint
+        )
+
+        try {
+          return await pRetry(
+            async () => {
+              await rateLimiter()
+              const request = await fetch(apiUrl.toString(), {
```

Comment:
Please can we rename this? `fetch` returns a [`Response`](https://developer.mozilla.org/en-US/docs/Web/API/Response) object and it's being stored in a variable named `request` 🤪

Note: more than one instance of this.